### ISSUE TITLE
[question] have find_package use an existing xxxConfig.cmake

### ISSUE THREAD

**Issue by myzb** [conan-io/conan#13116]
### What is your question?

Hi, I hope I can phrase my issue well enough. (tl;dr version at the bottom :D )

I'm currently developing an application that will cross compile from linux to qnx. I was previously using conan 1.54 and now upgraded to 1.58. While at it I also decided to move to host and build profiles. Previously I was only using 1 profile. So this brought the issue of some info no longer being available in `deps_cpp_info`

## My Setup / Workflow

This is an extract of how my conanfile.py currently looks like. I added some comments to explain a few things.
```
class AppConan(ConanFile):
# ...
   name = "MyApp"
   generators = "virtualenv", "CMakeDeps", "CMakeToolchain"
   requires = ["lz4/1.9.4"]
   tool_requires = ["qnx-compiler", "cmake-helpers"]

   def generate(self):
      deps = CMakeDeps(self)
      # I kinda want to use HelpersConfig.cmake provided
      # by the helpers package and not generate a new one
      # deps.build_context_activated = ["cmake-helpers"]
      deps.generate()

      tc = CMakeToolchain(self)
      # This was the proposal from the colleague that created the "cmake-helpers" 
      # package to allow finding the HelpersConfig.cmake shipped within his package
      # in conan 1.54
      for _, val in self.deps_cpp_info.dependencies:
         path = val.rootpath.replace("\\", "/")
         name = val.get_property("cmake_file_name")
         if name != None:
            tc.variables[f"{name}_ROOT"] = str(path)
     tc.generate()

# ...
```

In my CMakeLists.txt I have this part:
```
find_package(cmake-helpers 0.0.2 CONFIG REQUIRED)
include(folder-inside-cmake-helpers/cmake/custom_cmake_commands) # <-- this is a .cmake file
``` 

This is how it worked in conan 1.54 with only one profile:
1.  conanfile.py would set `Helpers_ROOT = path/to/helpers-package/in/conancache`
2. `find_package` in `CMakeLists.txt` would find the corresponding `HelpersConfig.cmake` thanks to `Helpers_ROOT` and source it
3. In `CMakeLists.txt` i could call `include` to (optionally) include some more .cmake files provided by the helpers package

The cmake helpers is basically a package that provides many wrappers to "easily" setup up stuff we need for development, code analysis, etc.

## The issue

So now with 2 profiles I'm running into the issue that the `deps_cpp_info` no longer contains info about the build requirements.
This means that I can no longer set `Helpers_ROOT` (using the for-loop in my conanfile) because my `self.deps_cpp_info.dependencies` no longer contains info about my helper package.

Using:
```
deps.build_context_activated = ["cmake-helpers"]
```

also doesn't help since I want to use the existing `HelpersConfig.cmake`, since that file sets a few CMake variables that allows calling `include` to include some other cmake files.

From the CMake error log I see that I somehow have to get conan to set the path to this build requirement in `CMAKE_PREFIX_PATH` or alternatively somehow set `Helpers_ROOT`. 

So the question is how to do this if I no longer have to info info in `deps_cpp_info` ?

## Tl;Dr

How can I make `find_package(XXX CONFIG)` use the `XXXConfig.cmake` provided by a package that is a build requirement when using host and build profiles. This only seems to work for `find_package(XXX MODULE)` if the package provides a `FindModule.cmake`.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @myzb 

Thanks for your question.

The  ``generate()`` methods was implemented from the origin to work on the ``self.dependencies`` object, which is more complete than the ``self.deps_cpp_info`` one. Please have a look to https://docs.conan.io/en/latest/reference/conanfile/dependencies.html

Basically:
- It allows to access information from both the build and host context
- It allows to access more information from dependencies, beyond ``cpp_info``
- It also allows to access information about the "requires" type to that dependency

All the existing new integrations in https://docs.conan.io/en/latest/reference/conanfile/tools.html already use the ``self.dependencies`` object to implement their functionality.


Please, have a look at that and let us know if this helps.

**Comment by myzb** [comment]
@memsharded thanks a lot for your quick reply! I totally missed there was a `.dependencies` Attribute. It seems it's not (yet) listed under [Attribute](https://docs.conan.io/en/1.58/reference/conanfile/attributes.html?highlight=attributes) and it did not occur to me to just scroll a bit further down and look under [Dependencies](https://docs.conan.io/en/1.58/reference/conanfile/dependencies.html#dependencies).

Anyway I just had to adjust my for-loop to this:
```
generate(self):
      # Sets <package>_ROOT variables to support find_package
      for _, dep in self.dependencies.build.items():
         tc.variables[f"{dep.ref.name}_ROOT"] = str(dep.package_folder)
```

I have a few final questions, well mostly what I wrote in my tl;dr.

I noticed that with `find_package(XXX MODULE)`, cmake will correctly search in the conan package folder for XXX. Obviously since the package I'm trying to find only provides a XXXConfig.cmake and not a FindXXX.cmake the search will fail. 
Because of this I have to use `find_package(XXX CONFIG)`. But with `CONFIG` I also have to do that "trick" with my for-loop to set the `XXX_ROOT` variable to support `find_package`.

1. For `CONFIG` it seems that conan does not set the supporting cmake paths to have cmake look in the conan package folders.  I think it's `CMAKE_PREFIX_PATH` what's missing. Is this intended?
2. Is there a way to have `find_package(XXX CONFIG)` succeed in my case without the hacky for-loop? Perhaps something is missing in the recipe of the package I'm trying to find?

**Comment by memsharded** [comment]
> Because of this I have to use find_package(XXX CONFIG). But with CONFIG I also have to do that "trick" with my for-loop to set the XXX_ROOT variable to support find_package.

This shouldn't be necessary. The test in https://github.com/conan-io/conan/blob/e0a8ee058bc8dc2d9811b8aeb6999f69aeb78d85/conans/test/functional/toolchains/cmake/test_cmaketoolchain_paths.py#L176 is implementing this use case (packaging your own)

Maybe the issue is that the recipe should define the:
```python
def package_info(self):
     self.cpp_info.builddirs = ["."]  # if the xxx-config.cmake is in the root of the package
```

So the ``CMakeToolchain`` generates the ``conan_toolchain.cmake`` that puts the package path in the ``CMAKE_PREFIX_PATH^`. Can you please try that?

**Comment by myzb** [comment]
Sorry for the late reply, I just now had the chance to look at this.

I tried doing: 
```python
def package_info(self):
     self.cpp_info.builddirs = ["."]  # if the xxx-config.cmake is in the root of the package
```

and now the conan_toolchain.cmake has the following entry (notice the dot "." at the end):
```cmake
# Definition of CMAKE_MODULE_PATH
# Explicitly defined "buildirs" of "build" context dependencies
list(PREPEND CMAKE_MODULE_PATH "/home/myzb/.conan/data/helper/0.0.2/me/testing/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/."

# ... snip ...

# Definition of CMAKE_PREFIX_PATH, CMAKE_XXXXX_PATH
# The Conan local "generators" folder, where this toolchain is saved.
list(PREPEND CMAKE_PREFIX_PATH ${CMAKE_CURRENT_LIST_DIR} )
```
I was expecting this to be added to the `list(PREPEND CMAKE_PREFIX_PATH ...) ` part above instead of `CMAKE_MODULE_PATH`

When I do 

```
find_package(helper 0.0.2 REQUIRED)
```

It will try to search for `Findhelper.cmake` under `CMAKE_MODULE_PATH`, but my helper packge only provides `helperConfig.cmake` so I get the following error:

```
CMake Error at CMakeLists.txt:11 (find_package):
  By not providing "Findhelper.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "helper", but
  CMake did not find one.

  Could not find a package configuration file provided by "helper" (requested
  version 0.0.2) with any of the following names:

    helperConfig.cmake
    helper-config.cmake

  Add the installation prefix of "helper" to CMAKE_PREFIX_PATH or set "helper_DIR"
  to a directory containing one of the above files.  If "helper" provides a
  separate development package or SDK, be sure it has been installed.
```

If I do the following instead (since it needs to find `helperConfig.cmake`):
```
find_package(helper 0.0.2 CONFIG REQUIRED)
```

I get the bottom part of the error above saying that I should be adding `helperConfig.cmake` to `CMAKE_PREFIX_PATH`. But as you can see in the snip from `conan_toolchain.cmake`  the path is not added to `CMAKE_PREFIX_PATH` but only `CMAKE_MODULE_PATH`.

I've been reading [cmake_file_name](https://docs.conan.io/1/reference/conanfile/tools/cmake/cmakedeps.html?highlight=cmake_file_name#properties), could it be that I need to set something extra from there. I'm having a hard time understanding these settings. The recipe for `helper` currently only has this:

```python
    def package(self):
        # Set CMake config file Name for CMakeDeps generator
        self.cpp_info.set_property("cmake_file_name", "helper")
```

Perhaps something extra is needed for conan to add the path to `CMAKE_PREFIX_PATH`?

I also looked at [test](https://github.com/conan-io/conan/blob/e0a8ee058bc8dc2d9811b8aeb6999f69aeb78d85/conans/test/functional/toolchains/cmake/test_cmaketoolchain_paths.py#L176) you posted. I'm not the python expert but if I understood the code correctly, that test is for module mode? and not config mode?

I hope you can somewhat follow my issue, if not I'll try to get some time to make a minimum working example that I can make available publicly. Else I'd have posted my actual stuff already :)

**Comment by myzb** [comment]
Could my issue form above be related to this Note?

> In Conan 1.45 the CMakeToolchain doesn’t append the root package folder of the dependencies (declared in the cpp_info.builddirs) to the CMAKE_PREFIX_PATH variable. That interfered with the find_file, find_path and find_program, making, for example, impossible to locate only the executables from the build context. In Conan 2.0, the cppinfo.builddirs won’t contain by default the '' entry (root package).

From: https://docs.conan.io/1/reference/conanfile/tools/cmake/cmaketoolchain.html#extending-and-customizing-cmaketoolchain

Thing is it's being added to CMAKE_MODULE_PATH ...

**Comment by myzb** [comment]
So my colleague found out what was going on.

Since our cmake-helper is a tool_require we need to additionally set `force_host_context=True` for our cmake-helper. And also add ~~`self.cpp_info.builddirs = ["."]`~~ `self.cpp_info.builddirs = ["helpers"]` to the cmake-helper recipe.

**Comment by memsharded** [comment]
Thanks for following up @myzb, happy that you found a solution, sorry for not being able to respond earlier.

However, I think that it might not be necessary to do the ``force_host_context=True``, that shouldn't be necessary (and in fact, counter-productive, as it doesn't work in Conan 2.0), so I am re-opening the issue.

I have realized that you might be missing the ``build_context_build_modules`` definition.
As build modules in the "build" context are problematic, they are also opt-in, so your recipe should be something like:

```python
def generate(self):
     deps = CMakeDeps(self)
     deps.build_context_activated = ["cmake-helpers"]
     deps.build_context_build_modules = ["cmake-helpers"]
     deps.generate()
```
Please try that and let us know.

**Comment by myzb** [comment]
> ```python
> def generate(self):
>      deps = CMakeDeps(self)
>      deps.build_context_activated = ["cmake-helpers"]
>      deps.build_context_build_modules = ["cmake-helpers"]
>      deps.generate()
> ```

We haven't switched to conan 2.0 yet but with 1.59 this does not work. But I think that what we are trying to do is no intended by conan.

Our idea was to have our `cmake-helper` package provide a custom `helperConfig.cmake` that is located in the package root. With that using `find_package()` would find and run the `helperConfig.cmake` located in the `cmake-helper` package root. The `cmake-helper` is a tool_requires that adds some subfolders to `CMAKE_MODULE_PATH` that contain cmake files which define some functions the user can `include` from his `CMakeLists.txt` if he desires.

From the 2nd note in [customizing cmake toolchain](https://docs.conan.io/1/reference/conanfile/tools/cmake/cmaketoolchain.html#extending-and-customizing-cmaketoolchain) it would seem that not appending the root path of packages to `CMAKE_PREFIX_PATH` is intended and has a valid reason. Atleast this is supported by [this](https://github.com/conan-io/conan/blob/1.59.0/conan/tools/cmake/toolchain/blocks.py#L467) and [this](https://github.com/conan-io/conan/blob/1.59.0/conan/tools/cmake/toolchain/blocks.py#L542).

So only package subfolders or "non-root" paths are getting added to `CMAKE_MODULE_PATH` and `CMAKE_PREFIX_PATH`. Root paths only get added to `CMAKE_MODULE_PATH` .

----

Long story short, we found a much simpler way to achieve what we want, basically:

1. cmake-helper recipe: add the helpers subfolder where the cmake files are
    ```python
    def package_info(self):
      self.cpp_info.builddirs = ["helpers"]
    ```
2. Consumer recipe: Remove `deps = CMakeDeps` stuf from `generate(self)`, the default via the attribute `generators = "CMakeDeps"` is enough.
3. Remove `find_package(cmake-helper CONFIG)` from our `CMakeLists.txt` <- not needed anymore
4. Just directly call `include(helper-script-1)` from `CMakeLists.txt` since `/path/to/.conan/data/cmake-helper-root/helpers` is in `CMAKE_MODULE_PATH`

**Comment by memsharded** [comment]
Sorry this was not followed up back then.

Now there is a full example of this, with existing code in https://docs.conan.io/2/examples/tools/cmake/cmake_toolchain/use_package_config_cmake.html. 

It also seems that you found a solution, so I am closing the ticket, but don't hesitate to comment, reopen or create a new ticket for any further question. Thanks for the feedback!
